transactions/sell_item.cdc

```diff
@@ -1,43 +1,75 @@
-import NFTStorefront from "../contracts/NFTStorefront.cdc"
+import FlowToken from 0x0ae53cb6e3f42a79
+import FungibleToken from "../contracts/utility/FungibleToken.cdc"
+import NonFungibleToken from "../contracts/utility/NonFungibleToken.cdc"
+import ExampleNFT from "../contracts/utility/ExampleNFT.cdc"
+import MetadataViews from "../contracts/utility/MetadataViews.cdc"
+import NFTStorefrontV2 from "../contracts/NFTStorefrontV2.cdc"
 
-transaction(saleItemID: UInt64, saleItemPrice: UFix64) {
-    let flowReceiver: Capability<&FlowToken.Vault{FungibleToken.Receiver}>
-    let exampleNFTProvider: Capability<&ExampleNFT.Collection{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>
-    let storefront: &NFTStorefront.Storefront
+transaction(saleItemID: UInt64, saleItemPrice: UFix64, customID: String?, commissionAmount: UFix64, expiry: UInt64, marketplacesAddress: [Address]) {
+    let flowReceiver: Capability<&AnyResource{FungibleToken.Receiver}>
+    let exampleNFTProvider: Capability<&AnyResource{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>
+    let storefront: &NFTStorefrontV2.Storefront
+    var saleCuts: [NFTStorefrontV2.SaleCut]
+    var marketplacesCap: [Capability<&AnyResource{FungibleToken.Receiver}>]
 
     prepare(acct: AuthAccount) {
+        self.saleCuts = []
+        self.marketplacesCap = []
         // We need a provider capability, but one is not provided by default so we create one if needed.
         let exampleNFTCollectionProviderPrivatePath = /private/exampleNFTCollectionProviderForNFTStorefront
 
-        self.flowReceiver = acct.getCapability<&FlowToken.Vault{FungibleToken.Receiver}>(/public/flowTokenReceiver)!
+        self.flowReceiver = acct.getCapability<&{FungibleToken.Receiver}>(/public/flowTokenReceiver)!
         assert(self.flowReceiver.borrow() != nil, message: "Missing or mis-typed FlowToken receiver")
 
-        if !acct.getCapability<&ExampleNFT.Collection{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath)!.check() {
-            acct.link<&ExampleNFT.Collection{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath, target: /storage/NFTCollection)
+        if !acct.getCapability<&{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath)!.check() {
+            acct.link<&{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath, target: ExampleNFT.CollectionStoragePath)
         }
 
-        self.exampleNFTProvider = acct.getCapability<&ExampleNFT.Collection{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath)!
+        self.exampleNFTProvider = acct.getCapability<&{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>(exampleNFTCollectionProviderPrivatePath)!
+        let collection = acct
```

Comment:
I see problems with this approach. What if my NFT changes the way royalties are handled between when I listed it and when I sell it? Then this code will use the old royalties information. 

I belive when we worked on Royalties we stated that Royalties should be calculated at the time of settlment, not at the time of listing.

Reply:
The easiest way to do this is to require that the capability also links Metadata.ViewResolver OR merge the default impl PR and default implement the metadata methods in NFT.CP

Reply:
The problem I see here with the dynamic royalty behaviour is that it will make it hard to calculate the static sale price of the listing. Although I envision minimal changes in the royalty details once the NFT get launched and if anyone does that then it would lead to changes in the sale price. 

Another problem I can foresee is if somehow the wallet (has the authority to change the royalty information) gets compromised then it can change the royalty cuts to 0 or make it too high so it can rug the seller.

Reply:
Have you considered not calculating the cuts as static parts up front but rather send them in as percentages?

Reply:
An important aspect that came up in a discussion today is that if we decided to calculate royalties at settlement time then we have en ensure the percentage is not increased since you listed it. That was a huge problem on ETH a while back.

Reply:
@bjartek So are you saying in your last comment that you are ok with calculating them up front? that is what we agree should be done